doc/developer/design/20230903_avro_doc.md

```diff
@@ -0,0 +1,388 @@
+  * If the key schema, a `KEY DOC ON TYPE` option specifically naming the field's
+    type. If the value schema, a `VALUE DOC ON TYPE` option specifically naming the
+    field's type.
+  * A `DOC ON TYPE` option specifically naming the field's type.
+  * A comment on the field's type.
+
+For example, consider the following sink:
+
+```sql
+CREATE TYPE point AS (x integer, y integer);
+CREATE MATERIALIZED VIEW v AS SELECT ROW(1, 1)::point AS c1, 'text' AS c2;
+CREATE SINK FROM v INTO KAFKA ... FORMAT AVRO ...;
+```
+
+For `c1`, Materialize will consider the following sources of documentation
+in order:
+
+  * A `KEY DOC ON COLUMN` or `VALUE DOC ON COLUMN` option naming `v.c1`.
+  * A `DOC ON COLUMN` option naming `v.c1`.
+  * A comment on column `v.c1`.
+  * A `KEY DOC ON TYPE` or `VALUE DOC ON TYPE` option naming `point`.
+  * A `DOC ON TYPE` option naming `point`.
+  * A comment on type `point`.
+
+`c2` is analogous.
+
+For `c1.x`, Materialize will consider the following sources of documentation
+in order:
+
+  * A `KEY DOC ON COLUMN` or `VALUE DOC ON COLUMN` option naming `point.x`.
+  * A `DOC ON COLUMN` option naming `point.x`.
+  * A comment on column `point.x`.
+  * A `KEY DOC ON TYPE` or `VALUE DOC ON TYPE` option naming `integer`.
+  * A `DOC ON TYPE` option naming `integer`.
+  * A comment on type `integer`.
+
+`c1.y` is analogous.
+
+Materialize will also use comments on the materialized view itself as the
+default documentation for the top-level record (named `row` for the key schema
+and `envelope` for the value schema). In the example above, Materialize will
+consider the following sources of documentation in order:
+
+  * A `DOC ON TYPE` comment naming `v`.
+  * A comment on materialized view `v`.
+
+Changes made to comments *after* the the sink is created will not be reflected
+in the Avro schema.
+
+### User-facing documentation
+
+The details in the [Design](#design) section above will need to be worked into
+the Avro sink documentation. In particular, we should be sure to precisely
+explain Avro schema node specifiers and the order in which Materialize searches
+for sources of documentation.
+
+Note that we expect the `COMMENT` interface to be the most common method by
+which users add comments to their sink schemas. The `DOC ON` option is primarily
+useful for the implementation, so that a `CREATE SINK` command can be fully self
+describing. We do expect users to use `DOC ON` in a pinch, e.g. if a single sink
+needs to customize the documentation for a nested type, rather than pulling its
+default description from its comment.
+
+### Implementation
+
+#### Parser
+
+The extensions to the parser are straightforward.
+
+#### Planner
+
+The planner will "freeze" any comments that have been promoted to documentation
+by rewriting the `CREATE SINK` statement to contain those comments as explicit
+`DOC ON` options. This ensures that the Avro schema generated for a sink is
```

Comment:
To clarify, all of the comments (including both user provided DOC ON and from comments) will be expanded to `KEY DOC ON` and `VALUE DOC ON` right? So the planner will essentially freeze an expanded version which will only contain `KEY DOC ON` and `VALUE DOC ON`.

Reply:
Oh, hadn't thought about it in that much depth. I don't think you strictly speaking _need_ to do that normalization, and can just reapply the priorities when you replan on every startup, but if makes the implementation simpler, go for it.